src/urllib3/connection.py

```diff
@@ -612,63 +663,104 @@ def set_cert(
         self.ca_cert_data = ca_cert_data
 
     def connect(self) -> None:
-        sock: socket.socket | ssl.SSLSocket
-        self.sock = sock = self._new_conn()
-        server_hostname: str = self.host
-        tls_in_tls = False
-
-        # Do we need to establish a tunnel?
-        if self._tunnel_host is not None:
-            # We're tunneling to an HTTPS origin so need to do TLS-in-TLS.
-            if self._tunnel_scheme == "https":
-                # _connect_tls_proxy will verify and assign proxy_is_verified
-                self.sock = sock = self._connect_tls_proxy(self.host, sock)
-                tls_in_tls = True
-            elif self._tunnel_scheme == "http":
-                self.proxy_is_verified = False
-
-            # If we're tunneling it means we're connected to our proxy.
-            self._has_connected_to_proxy = True
-
-            self._tunnel()  # type: ignore[attr-defined]
-            # Override the host with the one we're requesting data from.
-            server_hostname = self._tunnel_host
-
-        if self.server_hostname is not None:
-            server_hostname = self.server_hostname
-
-        is_time_off = datetime.date.today() < RECENT_DATE
-        if is_time_off:
-            warnings.warn(
-                (
-                    f"System time is way off (before {RECENT_DATE}). This will probably "
-                    "lead to SSL verification errors"
-                ),
-                SystemTimeWarning,
+        # Today we don't need to be doing this step before the /actual/ socket
+        # connection, however in the future we'll need to decide whether to
+        # create a new socket or re-use an existing "shared" socket as a part
+        # of the HTTP/2 handshake dance.
```

Comment:
So, thinking out loud, here are the cases we need to cover, bearing in mind this is the HTTPS connection.

1. If we don't support HTTP/2 (because it's not in `ALPN_PROTOCOLS`, like today), then we don't change anything. `connect()` calls should return connected sockets, without any blocking.
2. If we do support HTTP/2, there are three cases:
   1. We know if the target supports HTTP/2 or not. If it does not, it is equivalent to 1. above. If it does, we should use the existing shared socket.
   2. We don't know what the target supports, and no connection is ongoing. We should connect and prevent other connections to do so.
   3. We don't know what the target supports, and a connection is ongoing. We should wait for that connection to go through and then it becomes equivalent to 2.i.
3. With HTTP/2 prior knowledge, which is what `urllib3.http2.inject_into_urllib3` does that right now, we need to create a single connection too, like in 2.ii and 2.iii. We should do the equivalent of https://github.com/python-hyper/hyper/blob/b77e758f472f00b098481e3aa8651b0808524d84/hyper/http20/connection.py#L345-L347, even though it may be complicated as we need to support 1. and 2. too. We're not going to support this right away.

Given the above, is there a way to avoid the `None` value for the probe cache and replace by waiting on the `target_supports_http2` lock? I think this would simplify the three last lines of `_HTTP2ProbeCache.get` and avoid returning a value without holding for a lock. Maybe `_HTTP2ProbeCache.get(host, port)` could be used as a context manager even?

Reply:
Yes I like your suggested approach much better, thank you! Let me implement it that way.

Reply:
This comment was from 3 weeks ago so I'm a little hazy now, I think we decided to handle the socket sharing in a separate PR but to build on the `HTTP2ProbeCache` into a proper socket-sharing solution for HTTP/2

Reply:
Yes! This is resolved, thanks.